### PR TITLE
Feature/v2/sensor update UI

### PR DESCRIPTION
**Pull request by hage1005**
# Administrative Info
Make sure your branch name conforms to: `feature/hotfix/.../describe_your_branch_with_underscores

### Changes
What changes did you make?
- closing #52 
- 
## Checklist

Remove any items which are not applicable.

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tested my program thoroughly
- [x] Did you add/sub fields in the environment file? if so please update .env_sample

### Confirmation of Change 
Upload a screenshot, if possible. Otherwise, please provide instructions on how to see the change.

| [TODO Screenshot Title] |
| --- |
| [TODO Screenshot Link] |

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (mingkaili) [inline .vscode/launch.json:1]: remove IDE specific file, add ignore for all .vscode files in .gitignore

File: database/model/lotModel.js [hunk 2]
```diff
@@ -91,6 +91,7 @@ async function insertWithSpotsAndOwnership (parking_lot_info, spots_info, admin_
     const result = { insertResult: 'failed' };
     await db.transaction(async (transaction) => {
         try {
+            console.log("Im here 1");
```
Comment (mingkaili) [inline database/model/lotModel.js:94]: remove debug logs

File: database/model/spotsModel.js [hunk 3]
```diff
@@ -23,26 +23,73 @@ async function getSpots(lot_id) {
...
-async function update(spot_id, spot_info) {
-    try {
-        await db('spots')
-            .where({ spot_id })
-            .update(spot_info);
-    } catch (err) {
-        console.log(err);
-    }
+async function update (spot_hashes, spot_info) {
+    console.log("spot_hashes", spot_hashes);
```
Comment (mingkaili) [inline database/model/spotsModel.js:45]: remove debug logs

File: database/model/spotsModel.js [hunk 4]
```diff
@@ -23,26 +23,73 @@ async function getSpots(lot_id) {
...
+    console.log("spot_info", spot_info);
```
Comment (mingkaili) [inline database/model/spotsModel.js:46]: remove debug logs

File: database/model/spotsModel.js [hunk 5]
```diff
@@ -23,26 +23,73 @@ async function getSpots(lot_id) {
...
+    // Update the spots with the given spot_hashes
```
Comment (mingkaili) [inline database/model/spotsModel.js:47]: Remove unneeded comments

File: database/model/spotsModel.js [hunk 6]
```diff
@@ -23,26 +23,73 @@ async function getSpots(lot_id) {
...
+    //         .update(spot_info);
+    //     return { status: 'success' };
+    // } catch (err) {
+    //     console.log("Im here too");
+    //     return { status: 'failed', payload: err };
+    // }
+    const result = { status: 'failed' };
+    await db.transaction(async (trx) => {
+        try {
+            console.log("Hash: " + spot_hashes);
```
Comment (mingkaili) [inline database/model/spotsModel.js:61]: ditto

File: database/model/spotsModel.js [hunk 7]
```diff
@@ -23,26 +23,73 @@ async function getSpots(lot_id) {
...
+            console.log(spot_info);
```
Comment (mingkaili) [inline database/model/spotsModel.js:62]: ditto

File: database/model/spotsModel.js [hunk 8]
```diff
@@ -23,26 +23,73 @@ async function getSpots(lot_id) {
...
+                    .transacting(trx);
+            }
+            await trx.commit();
+            result.status = 'success';
+        } catch (err) {
+            await trx.rollback();
+        }
+    });
+
+    // try {
```
Comment (mingkaili) [inline database/model/spotsModel.js:76]: ditto

File: routes/firmware.js [hunk 9]
```diff
@@ -0,0 +1,140 @@
+var express = require('express');
```
Comment (mingkaili) [inline routes/firmware.js:1]: const

File: routes/firmware.js [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+var router = express.Router();
```
Comment (mingkaili) [inline routes/firmware.js:2]: ditto

File: routes/firmware.js [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+const s3_service = require('../services/s3-upload');
+const spotsModel = require('../database/model/spotsModel');
+
+router.get('/', async function (req, res, next) {
+    try {
+        const firmwareControl = await firmwareControlModel.get();
+        const data = await firmwareVersionModel.getWithNumSpots();
+        res.render('page/firmwareControl/firmwareControl.ejs', { title:"ProkoPark - Firmware Control", firmwareControl, data, moment });
+    } catch (err){
+        res.send(err)
```
Comment (mingkaili) [inline routes/firmware.js:20]: nit: instead of using send. use a status code instead: 500

File: routes/firmware.js [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+ * @param {file} req.body.ESP32
+ * @param {file} req.body.ESP8266
+ */
+router.post('/upload', upload.fields([{ name: 'ESP32' },{ name: 'ESP8266' }]), async function (req, res, next) {
+    try {
+        const { ESP32, ESP8266 } = req.files;
+        if (ESP32 == undefined || ESP8266 == undefined || req.body.version == '') {
+            return res.send('All fields are required');
+        }
+        if (await firmwareVersionModel.exists(req.body.version)) {
```
Comment (mingkaili) [inline routes/firmware.js:36]: Bad structure: never put functions in the condition of if statements.

```
const firmwareVersions = firmwareVersionModel.getByVersion(req.body.version);
const isFirmwareExists = firmwareVersions.length !== 0;
if (!isFirmwareExists) {
  blah blah
}
```

File: routes/firmware.js [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+            return res.send('Version already exists');
```
Comment (mingkaili) [inline routes/firmware.js:37]: status code: 401 or 409

File: routes/firmware.js [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+ * Upload new firmware version
+ * @param {string} req.body.firmware_version
+ * @param {file} req.body.ESP32
+ * @param {file} req.body.ESP8266
+ */
+router.post('/upload', upload.fields([{ name: 'ESP32' },{ name: 'ESP8266' }]), async function (req, res, next) {
+    try {
+        const { ESP32, ESP8266 } = req.files;
+        if (ESP32 == undefined || ESP8266 == undefined || req.body.version == '') {
+            return res.send('All fields are required');
```
Comment (mingkaili) [inline routes/firmware.js:34]: status code: 400

File: routes/firmware.js [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+    }
+    
+});
+
+/* Download ESP8266 firmware */
+router.get('/download/ESP8266/:version', async function (req, res, next) {
+    try {
+        const version = req.params.version;
+        const data = await firmwareVersionModel.getByVersion(version);
+        return res.redirect(data[0].ESP8266_url);
```
Comment (mingkaili) [inline routes/firmware.js:71]: if you are going to user return, make sure you return on every res operation to keep them consistent.

File: routes/firmware.js [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+        const version = req.params.version;
+        const data = await firmwareVersionModel.getByVersion(version);
+        return res.redirect(data[0].ESP8266_url);  
+    } catch (err) {
+        res.status(500).json({ message: "Error downloading ESP8266 firmware", error: err })
+    }  
+});
+
+/* Download ESP32 firmware */
+router.get('/download/ESP32/:version', async function (req, res, next) {
```
Comment (mingkaili) [inline routes/firmware.js:78]: You are writing this route in the wrong place. There is a separate repository that already have those routes. This is why it's important to read the documentation and have a clear documentation instead of doing this too quickly.

File: routes/index.js [hunk 17]
```diff
@@ -2,6 +2,7 @@ var express = require('express');
 var router = express.Router();
 
 const lotsModel = require('../database/model/lotModel');
+const spotsModel = require('../database/model/spotsModel');
```
Comment (mingkaili) [inline routes/index.js:5]: why adding this?

**Review by mingkaili — CHANGES_REQUESTED** [review]
TL;DR:
1. Break down the feature to submit smaller PRs. Right now what we have here is a cluster of work without enough quality.
2. Read design docs thoroughly.
3. Read over the PR before requesting for reviews as there are lots of development specific things like console logs, comments that isn't needed in production.

This is an extremely big PR with 40 file changes and 6000 lines of changes. You have done the following things in this PR:
- Your linter or IDE probably automatically did a codebase wide refactor that inserted a bunch of whitespaces and a bunch of indentations. If you are going to do that, make a new branch called refactor and do it there bc these things are extremely misleading and takes spaces for reviewers to focus on the actual changes
- Full Frontend changes
- Inserted new routes
- Made new Model functions

The above 4 things can be separated to at least 4 different PRs for readability enhancement.

**Keep In Mind** You always want to submit small size PRs in the industry due to: 
1. Small PRs gets faster code reviews
4. Small PRs get better reviews with a smaller size
5. Small PRs SIGNIFICANTLY improve code quality(You did not read over your code change bc you have lots of left over console log statements, and unneeded comments that makes the PR dirty).
6. Small PRs help protect the flow
There are many more reasons you can refer to [this simple article](https://www.swarmia.com/blog/why-small-pull-requests-are-better/) for the benefits

I only reviewed the backend portion of your PR and due to it's size, and it was a quick skim through. You need to read over the PR before asking for review to remove some of the quick things like: unneeded comments, console logs statements, etc.

**Comment by mingkaili** [comment]
The lint PR was merged. Make sure to rebase this branch.

File: routes/firmware.js (hunk 16 above)
Comment (hage1005) [reply routes/firmware.js:78]: Do you refer to the sensor-version-control repo?

**Comment by mingkaili** [comment]
We can close this PR now right?
